level00/LawPropertiesBonus.hs

```diff
@@ -0,0 +1,49 @@
+{-# OPTIONS_GHC -Wno-unused-binds -Wno-unused-imports #-}
+{-# LANGUAGE RankNTypes #-}
+module LawPropertiesBonus 
+  ( myBTreePrismLaws
+  , firstPrismLaw
+  , secondPrismLaw
+  , thirdPrismLaw
+  , _Empty
+  , _Node
+  ) where
+
+import           Control.Lens   (Prism', matching, preview, prism, review)
+
+import           Hedgehog
+import qualified Hedgehog.Gen   as Gen
+import qualified Hedgehog.Range as Range
+
+import           MyBTree        (MyBTree (..))
+
+-- Complete the following prisms, then write properties that check the Prism laws hold.
+_Empty :: Prism' (MyBTree k a) ()
+_Empty = error "_Empty prism not implemented"
+
+_Node :: Prism' (MyBTree k a) (MyBTree k a, (k, a), MyBTree k a)
+_Node = error "_Node prism not implemented"
+
+-- First, if I re or review a value with a Prism and then preview or use (^?),
+-- I will get it back:
+--
+-- preview l (review l b) ≡ Just b
+--
+firstPrismLaw :: (Eq b, Show b) => Gen b -> Prism' a b -> Property
+firstPrismLaw _gen _p = error "firstPrismLaw not implemented"
+
+-- Second, if you can extract a value a using a Prism l from a value s, then
```

Comment:
A prism must obey functions given to it, except where such functions would conflict with the First law.